spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala

```diff
@@ -127,545 +124,15 @@ class BatchNormalization[T: ClassTag](
-          (ev.toType[Float](runningMean.valueAt(_f)),
-            1 / Math.sqrt(ev.toType[Float](runningVar.valueAt(_f)) + eps).toFloat)
-        }
-
-        var sum = 0.0f
-        var i = 0
-        while (i < n) {
-          val index = i % gradOutputStride2 + (_f - 1) * gradOutputStride2 + gradOutputOffset +
-            (i / gradOutputStride2) * gradOutputStride
-          sum += gradOutput(index)
-          i += 1
-        }
-
-        var dotp = 0.0f
-        i = 0
-        while (i < n) {
-          val inputIndex = i % inputStride2 + (_f - 1) * inputStride2 + inputOffset +
-            (i / inputStride2) * inputStride
-          val gradOutputIndex = i % gradOutputStride2 + (_f - 1) * gradOutputStride2 +
-            gradOutputOffset + (i / gradOutputStride2) * gradOutputStride
-          dotp += (input(inputIndex) - mean) * gradOutput(gradOutputIndex)
-          i += 1
-        }
-
-        if (null != gradInput) {
-          if (train) {
-            val k = dotp * invstd * invstd / n
-            i = 0
-            while (i < n) {
-              val inputIndex = i % inputStride2 + (_f - 1) * inputStride2 + inputOffset +
-                (i / inputStride2) * inputStride
-              val gradInputIndex = i % gradInputStride2 + (_f - 1) * gradInputStride2 +
-                gradInputOffset + (i / gradInputStride2) * gradInputStride
-              gradInput(gradInputIndex) = (input(inputIndex) - mean) * k
-              i += 1
-            }
-
-            val gradMean = sum / n
-            i = 0
-            while (i < n) {
-              val gradInputIndex = i % gradInputStride2 + (_f - 1) * gradInputStride2 +
-                gradInputOffset + (i / gradInputStride2) * gradInputStride
-              val gradOutputIndex = i % gradOutputStride2 + (_f - 1) * gradOutputStride2 +
-                gradOutputOffset + (i / gradOutputStride2) * gradOutputStride
-              gradInput(gradInputIndex) = (gradOutput(gradOutputIndex) - gradMean -
-                gradInput(gradInputIndex)) * invstd * w
-              i += 1
-            }
-          } else {
-            var i = 0
-            while (i < n) {
-              val gradInputIndex = i % gradInputStride2 + (_f - 1) * gradInputStride2 +
-                gradInputOffset + (i / gradInputStride2) * gradInputStride
-              val gradOutputIndex = i % gradOutputStride2 + (_f - 1) * gradOutputStride2 +
-                gradOutputOffset + (i / gradOutputStride2) * gradOutputStride
-              gradInput(gradInputIndex) = gradOutput(gradOutputIndex) * invstd * w
-              i += 1
-            }
-          }
-        }
-
-        if (null != gradWeight && scaleW != 0) {
-          gradWeight(_f - 1 + gradWeightOffset) += scaleW * dotp * invstd
-        }
-
-        if (null != gradBias && scaleB != 0) {
-          gradBias(_f - 1 + gradBiasOffset) += scaleB * sum
-        }
-      })
-      f += 1
-    }
-    Engine.model.sync(results)
-  }
+  protected val gMean = Tensor[T]()
```

Comment:
override clearState() ?

Reply:
I can see in the original impl there's no clearState(). Not sure if we should clear the runningMean and runningVar as it is useful in the prediction

Reply:
clean gMean and gxMean buffer